### PR TITLE
Add support for dynamic pruning to cardinality aggregations on low-cardinality keyword fields.

### PR DESCRIPTION
**Pull request by jpountz**
On low-cardinality keyword fields, the `cardinality` aggregation currently uses the `global_ordinals` execution mode most of the time, which consists of collecting all documents that match the query, reading ordinals of the values that these documents contain, and setting bits in a bitset for these ordinals.

This commit introduces a feedback loop between the query and the `cardinality` aggregator, which allows the query to skip documents that only contain values that have already been seen by the `cardinality` aggregator. On the `nyc_taxis` dataset, a `match_all` query and the `vendor_id` field (2 unique values), the `cardinality` aggregation went from 3s to 3ms. The speedup would certainly not be as good in all cases, but I would still expect it to be very significant in many cases.

### CODE REVIEW

**Comment by jpountz** [comment]
I've run some benchmarks on a synthetically generated index that has 100M documents with the following values for the i-th indexed document:
 - `k0` is `i % 1000 < 50`
 - `k1` is `i % 500 < 25`
 - `k2` is `i % 500`

Then the following queries:

`match_all-cardinality`:

```json
      {
        "size": 0,
        "track_total_hits": false,
        "aggs": {
          "k2_cardinality": {
            "cardinality": {
              "field": "k2",
              "execution_hint": "global_ordinals"
            }
          }
        }
      }
```

`k0-cardinality`. This runs a cardinality aggregation on a `term` query that would collect every unique value of `k2` so it should be able to exit early. This is expected to simulate a cardinality on the number of agents sending data over a date range, since all agents are expected to exist given a sufficiently long time range.

```
      {
        "size": 0,
        "track_total_hits": false,
        "query": {
          "term": {
            "k0": "0"
          }
        },
        "aggs": {
          "k2_cardinality": {
            "cardinality": {
              "field": "k2",
              "execution_hint": "global_ordinals"
            }
          }
        }
      }
```

`k1-cardinality`. This runs a cardinality aggregation on a `term` query that does NOT see every value of `k2`. So it can prune some hits, but it cannot end early, it need to go through the whole index. This would typically happen when there is correlation between the query and the aggregated field, e.g. compute the cardinality of agents that reported errors: probably that some monitored hosts are more likely to have errors than others.

```json
      {
        "size": 0,
        "track_total_hits": false,
        "query": {
          "term": {
            "k1": "0"
          }
        },
        "aggs": {
          "k2_cardinality": {
            "cardinality": {
              "field": "k2",
              "execution_hint": "global_ordinals"
            }
          }
        }
      }
```

Here is what Rally reported:

```
|                                                Min Throughput | match_all-cardinality |    0.57575     | 444.294       |   443.718   |  ops/s | +77067.82% |
|                                               Mean Throughput | match_all-cardinality |    0.578541    | 444.294       |   443.715   |  ops/s | +76695.54% |
|                                             Median Throughput | match_all-cardinality |    0.57887     | 444.294       |   443.715   |  ops/s | +76651.97% |
|                                                Max Throughput | match_all-cardinality |    0.580407    | 444.294       |   443.713   |  ops/s | +76448.66% |
|                                       50th percentile latency | match_all-cardinality | 1719.91        |   1.88433     | -1718.02    |     ms |    -99.89% |
|                                       90th percentile latency | match_all-cardinality | 1753.1         |   2.1355      | -1750.96    |     ms |    -99.88% |
|                                      100th percentile latency | match_all-cardinality | 1791.7         |   2.63628     | -1789.06    |     ms |    -99.85% |
|                                  50th percentile service time | match_all-cardinality | 1719.91        |   1.88433     | -1718.02    |     ms |    -99.89% |
|                                  90th percentile service time | match_all-cardinality | 1753.1         |   2.1355      | -1750.96    |     ms |    -99.88% |
|                                 100th percentile service time | match_all-cardinality | 1791.7         |   2.63628     | -1789.06    |     ms |    -99.85% |
|                                                    error rate | match_all-cardinality |    0           |   0           |     0       |      % |      0.00% |
|                                                Min Throughput |        k0-cardinality |    0.5367      | 473.317       |   472.78    |  ops/s | +88090.11% |
|                                               Mean Throughput |        k0-cardinality |    0.538432    | 473.317       |   472.778   |  ops/s | +87806.55% |
|                                             Median Throughput |        k0-cardinality |    0.538581    | 473.317       |   472.778   |  ops/s | +87782.16% |
|                                                Max Throughput |        k0-cardinality |    0.539811    | 473.317       |   472.777   |  ops/s | +87581.97% |
|                                       50th percentile latency |        k0-cardinality | 1846.59        |   1.83855     | -1844.75    |     ms |    -99.90% |
|                                       90th percentile latency |        k0-cardinality | 1891.9         |   2.06683     | -1889.83    |     ms |    -99.89% |
|                                      100th percentile latency |        k0-cardinality | 1953.52        |   2.39496     | -1951.12    |     ms |    -99.88% |
|                                  50th percentile service time |        k0-cardinality | 1846.59        |   1.83855     | -1844.75    |     ms |    -99.90% |
|                                  90th percentile service time |        k0-cardinality | 1891.9         |   2.06683     | -1889.83    |     ms |    -99.89% |
|                                 100th percentile service time |        k0-cardinality | 1953.52        |   2.39496     | -1951.12    |     ms |    -99.88% |
|                                                    error rate |        k0-cardinality |    0           |   0           |     0       |      % |      0.00% |
|                                                Min Throughput |        k1-cardinality |    0.529033    |   1.876       |     1.34697 |  ops/s |   +254.61% |
|                                               Mean Throughput |        k1-cardinality |    0.532735    |   1.88155     |     1.34881 |  ops/s |   +253.19% |
|                                             Median Throughput |        k1-cardinality |    0.533744    |   1.8789      |     1.34515 |  ops/s |   +252.02% |
|                                                Max Throughput |        k1-cardinality |    0.534619    |   1.89503     |     1.36041 |  ops/s |   +254.46% |
|                                       50th percentile latency |        k1-cardinality | 1871.84        | 535.195       | -1336.64    |     ms |    -71.41% |
|                                       90th percentile latency |        k1-cardinality | 1901.33        | 539.858       | -1361.47    |     ms |    -71.61% |
|                                      100th percentile latency |        k1-cardinality | 1945.91        | 542.866       | -1403.04    |     ms |    -72.10% |
|                                  50th percentile service time |        k1-cardinality | 1871.84        | 535.195       | -1336.64    |     ms |    -71.41% |
|                                  90th percentile service time |        k1-cardinality | 1901.33        | 539.858       | -1361.47    |     ms |    -71.61% |
|                                 100th percentile service time |        k1-cardinality | 1945.91        | 542.866       | -1403.04    |     ms |    -72.10% |
|                                                    error rate |        k1-cardinality |    0           |   0           |     0       |      % |      0.00% |
```

The query on `k1` is between 3x and 4x faster. And the query on a `match_all` and on `k0` are ~1000x faster.

I also pushed some more tests that make sure that the optimization kicks in by using the debug info, so I think it's ready for someone to have a closer look at how it works.

**Comment by martijnvg** [comment]
@elasticmachine update branch

**Review by not-napoleon — APPROVED** [review]
Sorry it's taken me so long to look at this.  I think this is fine to merge.

**Comment by martijnvg** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/GlobalOrdCardinalityAggregator.java [hunk 1]
```diff
@@ -64,12 +90,182 @@ public GlobalOrdCardinalityAggregator(
...
     public LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, final LeafBucketCollector sub) throws IOException {
         values = valuesSource.globalOrdinalsValues(aggCtx.getLeafReaderContext());
+
+        if (parent == null && field != null) {
+            // This optimization only applies to top-level cardinality aggregations that apply to fields indexed with an inverted index.
+            final Terms indexTerms = aggCtx.getLeafReaderContext().reader().terms(field);
+            if (indexTerms != null) {
+                BitArray bits = visitedOrds.get(0);
+                final int numNonVisitedOrds = maxOrd - (bits == null ? 0 : (int) bits.cardinality());
+                if (maxOrd <= MAX_FIELD_CARDINALITY_FOR_DYNAMIC_PRUNING || numNonVisitedOrds <= MAX_TERMS_FOR_DYNAMIC_PRUNING) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/GlobalOrdCardinalityAggregator.java:211]: So the optimisation also kicks in on fields with more than 1024 unique values, if there 128 or less terms to be processed. The brute force leaf bucket collector implementation does update the bit sets used here to determine `numNonVisitedOrds`.

**Review by martijnvg — APPROVED** [review]
I think we should get this change merged. LGTM

**Pull request merged**